### PR TITLE
Custom Arrow Tips

### PR DESCRIPTION
**Pull request by behackl**
## List of Changes
- Refactor `ArrowTip` in `geometry.py` to be a (actually quite concrete) abstract base class so that different arrow tip shapes can be implemented by inheriting from it.
- Add `ArrowTriangleTip` (the current default tip), and two new tips `ArrowCircleTip`, `ArrowSquareTip`.
- Add keyword `tip_shape` to pass the custom shape to the `Arrow` constructor.

## Motivation
This is part of #272.

## Explanation for Changes
This PR changes the class hierarchy for arrow tips a little in order to allow the (easy) implementation of custom tips. Details are given in the List of Changes.

## Testing Status

As an example, rendering the `ArrowTest` scene in the Snippet

```python
import numpy as np

from manim import Scene, Arrow, ShowCreation, FadeOut,\
                  ArrowTriangleTip, ArrowSquareTip, ArrowCircleTip


class ArrowTest(Scene):
    def construct(self):
        a11 = Arrow(np.array([-2,3,0]), np.array([2, 3, 0]))
        a12 = Arrow(np.array([-2,2,0]), np.array([2, 2, 0]),
                    tip_shape=ArrowTriangleTip,
                    tip_style={'fill_opacity': 0, 'stroke_width': 3})
        a21 = Arrow(np.array([-2,1,0]), np.array([2, 1, 0]), tip_shape=ArrowSquareTip)
        a22 = Arrow(np.array([-2,0,0]), np.array([2, 0, 0]), tip_shape=ArrowSquareTip,
                    tip_style={'fill_opacity': 0, 'stroke_width': 3})
        a31 = Arrow(np.array([-2,-1,0]), np.array([2, -1, 0]), tip_shape=ArrowCircleTip)
        a32 = Arrow(np.array([-2,-2,0]), np.array([2, -2, 0]), tip_shape=ArrowCircleTip,
                    tip_style={'fill_opacity': 0, 'stroke_width': 3})
        self.play(ShowCreation(a11), ShowCreation(a12), ShowCreation(a21), 
                  ShowCreation(a22), ShowCreation(a31), ShowCreation(a32))
        self.wait(1)
        self.play(*[FadeOut(obj) for obj in self.mobjects])
```

yields

ArrowTest

## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

### CODE REVIEW

**Comment by huguesdevimeux** [comment]
Thanks for your contribution ! 
As this PR is still WIP; I'm marking it as draft.

**Comment by leotrs** [comment]
WOWOWOW This is beautiful thanks so much! I'll review in a sec, but thank you!

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:811]: Comment: I'm usually leery of multiple inheritance, but seeing as these classes are (a) very simple and (b) not meant to be inherited from, then I _guess_ it's ok.

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:808]: Suggestion: I'd rather all of these be properties

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:794]: Suggestion: perhaps add
```python
def __init__(self):
    pass
```
Explicity is better than implicit and whatnot

**Comment by behackl** [comment]
Thank you for this lightning-fast review! I was just typing that you don't need to stress yourself, as I don't know how soon I can add documentation and tests. 😄 

I'll look at your comments a bit later -- thanks again!

**Comment by huguesdevimeux** [comment]
Speaking of tests, I think we should wait for #340 to be merged before adding new tests, so it will be easier.

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py:792]: I fully agree. A more agressive suggestion is in bb1746d0b95510cf61e6b4bfe446ef1c55854738 (because I think it is better to avoid even the slightest chance of silent failures).

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: Agreed. I changed `base`, `tip_point` and `vector` in 38638475195b9940561deb1f49404e61dafba19a. Actually, `get_length` is not used anywhere -- and trying to refactor it as a `length` property would require additional work as the very same `length` is also passed via a corresponding entry in the `CONFIG` dict of `ArrowTip` (from where it is subsequently used to construct the tips...). I've removed the superfluous `get_length` for now.

Finally, changing `get_angle` to an `angle`-property reveals another problem: a `ArrowCircleTip` inherits from `Circle` which inherits from `Arc` -- and an `Arc` also has an `angle` property. I've renamed the former to `tip_angle`, see a01dd7f1f963b44651136d8a4982b85029834e4e.

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: 79757610679448cc0560ccffc9578d0254b5133c

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: 79757610679448cc0560ccffc9578d0254b5133c

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py:813]: I'm with you on being cautious regarding multiple inheritance. In this case, on the one hand, I believe it is actually semantically correct -- and on the other hand, I needed it for `ArrowCircleTip`: without having this class inherit from `Circle`, it would not populate `self.points`.

(But sure, I guess if it were necessary to remove multiple inheritance, a workaround could be found.)

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: I don't see a nice way to specify `ArrowTriangleTip` directly as the default value for `TipableVMobject.add_tip` (without some serious reordering I always get a `NameError` as `ArrowTriangleTip` is only defined later in the file -- and moving the implementation of the arrow tips before `TipableVMobject` is not possible because of `ArrowCircleTip`, which is a `Circle`, which is an `Arc`, which is a `TipableVMobject` again.)

Maybe this would work if the arrow tips were moved to a separate file -- but I think this might be overkill. I suggest simply handling the case of `tip_shape is None` (as it was before 7975761). Fixed via 249c2841.

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: I did this for the other arrow types as well; with 151d0a32 also start+end tip of double arrows and tips of curved arrows are customizable. :-)

**Comment by behackl** [comment]
> One more suggestion: I think the current way of creating filled/hollow tips is a bit awkward. Why not add a filled class to each? In this way this code
> 
> ```python
> a12 = Arrow(np.array([-2,2,0]), np.array([2, 2, 0]),
>                     tip_shape=ArrowTriangleTip,
>                     tip_style={'fill_opacity': 0, 'stroke_width': 3})
> ```
> 
> becomes this
> 
> ```python
> a12 = Arrow(np.array([-2,2,0]), np.array([2, 2, 0]),
>                     tip_shape=ArrowTriangleFilledTip)
> ```

Great idea! It really was rather clunky before. I've added some convenience classes for filled/hollow tips in e8d076ca. :-)

File: manim/mobject/geometry.py
Comment (leotrs) [reply manim/mobject/geometry.py]: Perfect, agreed on all accounts

File: manim/mobject/geometry.py
Comment (leotrs) [reply manim/mobject/geometry.py:834]: Yeah, agreed. Perhaps we can just leave a line somewhere in the docs recommending that people not inherit from them.

File: manim/mobject/geometry.py
Comment (leotrs) [reply manim/mobject/geometry.py:804]: Oh I like your solution. I was about to say this should be an AbstractBaseClass, but that would have meant triple inheritance which I'm not going anywhere near.

File: manim/mobject/geometry.py
Comment (leotrs) [reply manim/mobject/geometry.py]: CONFIG dicts are going away soon, so we can perhaps revisit this when that happens. This is fine for now

**Comment by leotrs** [comment]
Great great stuff. BTW, it was _extremely_ helpful of you to leave the commit link in each of your comments. Super easy to review!

**Comment by behackl** [comment]
> Great great stuff. BTW, it was _extremely_ helpful of you to leave the commit link in each of your comments. Super easy to review!

Thank you for the kind words and your reviews! The thing with the commit hashes in the comments is what I like most when I have to review stuff, glad you like it as well. 😄 

I'll finish this here up by writing some docstrings, then the draft status of the PR can be removed again. As @huguesdevimeux suggested above, I'll wait with adding tests until #335 is ready (or take care of that in a follow-up PR). 👍

**Comment by leotrs** [comment]
Sounds like a plan

**Comment by behackl** [comment]
I have made all changes I wanted to make.

- 80a07d5 simplifies the `Arrow.scale` method, which scales arrows while leaving arrow tips unchanged.
- 7e2ddf8 adds a basic form of documentation. As nobody had any serious concerns on Discord, I also added some explicit doctests (which are not performed yet in the pipeline, but I would like to open a separate PR to enable just that).
- a112cd3 is just an application of `black` to `geometry.py`, and
- 78adf579 merges the current master in here (because I hoped to get rid of the format test failure).

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:990]: Suggestion: Maybe add an Examples section that says something along the lines of "using ArrowFilledTip is a shorthand to using ArrowTip(...)`.

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:972]: Question: why not specify `.. code-block::` instead?

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:937]: Suggestion: "The vector with tail at the base of the arrow, and head at the tip of the arrow." (The word "span" has a precise meaning in linear algebra that may be confusing to some of our mathematically-adept readers (such as myself))

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:881]: Suggestion: "Instead, use one of the pre-defined ones, or make a custom one like this:"

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py]: Suggestion: add a "See also" section that points to all the pre-defined classes

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:664]: Suggestion: I think this behavior should be the default, but it should be optional. Can we add a boolean that would allow the user to propagate the scale() call to the tips?

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py:972]: AFAIK, it should not be necessary to do so, see https://www.sphinx-doc.org/en/1.5/markup/code.html -- but I admit, I did not try to actually generally the sphinx documentation.

Comment (leotrs) [reply manim/mobject/geometry.py:972]: I think you're right, it's not necessary, unless you want syntax highlighting. I think we want syntax highlighting :) Please use ` code-block:: python`

Comment (behackl) [reply manim/mobject/geometry.py:972]: I can add it, sure -- but the link I posted says that, as python is the default highlight language, it is sufficient to put `::` and still get syntax highlighting. :-)

(I'll try it later, currently I am not home.)

Comment (behackl) [reply manim/mobject/geometry.py:972]: https://manim.hackl.space/reference/manim.mobject.geometry.ArrowTip.html -- Syntax highlighting works indeed, but the formatting is off. I'll investigate and try to fix it.

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: Good call! cfb2cdaab5909b58859bc769cb914693b6bb951a

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: List added in 5528cc30a351ff57a5b06f3ab085620148cc2fe8

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: Reworded in 5528cc30a351ff57a5b06f3ab085620148cc2fe8

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py]: While I don't really see the danger of interpreting this as the span in a vector space sense, especially given that it is used as a verb with specified start and end point, I don't mind rewording. I wanted to avoid tail and head though, so I propose a construction using "pointing" in 5528cc30a351ff57a5b06f3ab085620148cc2fe8.

File: manim/mobject/geometry.py
Comment (behackl) [reply manim/mobject/geometry.py:1009]: I like that! 5528cc30a351ff57a5b06f3ab085620148cc2fe8

**Comment by behackl** [comment]
Formatting inconsitencies are fixed in e0acdeb8.

Changing how tips are added to arrows also slightly changed the visuals of `DoubleArrow`, which is why the corresponding doctest has to be fixed in 48acf99c.
Before: 
image

After: 
image

(Notice that the tips were not equally large before, this is fixed now.)

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:669]: Hmmm Line is a Vmobject, no? So it would be better to link to VMobject.scale instead. Why did you change it?

Comment (behackl) [reply manim/mobject/geometry.py:669]: There is no separate `VMobject.scale`, it is directly inherited from `Mobject.scale`, which caused the "See Also" block in the rendered HTML documentation just to include `scale()` without link (and in particular, without reference to `VMobject` or `Mobject`).

Comment (leotrs) [reply manim/mobject/geometry.py:669]: Ah good point

**Comment by leotrs** [comment]
@behackl can you please solve the conflicts and make sure that the tests still pass? We just merged #335 which completely rehauled the test suite

**Comment by behackl** [comment]
There was no actual conflict, I only needed to merge master locally. And of course, black still fails.

Unfortunately, the tests revealed another inconvenience: due to different architectures, numerics are slightly different between operating systems, which leads to numpy yielding `-0.0` in macos and `0.0` otherwise. 🙄

I'll come up with a fix for this, or replace the faulty doctests.

**Comment by behackl** [comment]
- Fixed doctests via 5b6f3aa6
- And I also added a graphical unit test in 9f29fdfe -- this was very painless and straightforward, thanks to #335  :-)

Ready to be reviewed again!

File: manim/mobject/geometry.py
Comment (leotrs) [inline manim/mobject/geometry.py:935]: lol, I did not know about this hack

**Comment by leotrs** [comment]
Can we get one more approval here? This is super ready @huguesdevimeux @Aathish04

**Review by Aathish04 — APPROVED** [review]
Thank you for the contributions!

**Review by huguesdevimeux — APPROVED** [review]
SGTM !

**Comment by leotrs** [comment]
:partying_face:

**Pull request merged**
